ASMDocumentation.md

```diff
@@ -0,0 +1,5841 @@
+# Documenting the Inline Assembly
+
+Thanks to [xezon](https://github.com/TheSuperHackers/GeneralsGameCode/pull/405#issuecomment-2709737713) for suggesting
+using [godbolt](https://godbolt.org/) to compare assembly outputs.
+
+All generated code uses the latest version of MSVC 19 x86, as available in godbolt.
+
+As recommended by [tomson26](https://github.com/TheSuperHackers/GeneralsGameCode/pull/405#issuecomment-2710604104), I am
+not including `blitblit`, `rlerle`, `lcw` or `mpu`.
+
+The `crc` assembly block is already extensively documented in source.
+
+> **NOTE**: For methods with `__forceinline`, this is removed in godbolt to allow the compiler to only emit the function
+> assembly and not other, unrelated assembly to keep the function from being removed.
+
+> **NOTE**: All the code is compiled **without** optimization. Optimized code will result in smaller and more optimized
+> assemblies in modern compilers.
+
+> **NOTE**: Classes and structures, as well as functions that don't exist, will be replicated in godbolt in their most
+> basic necessity to compile, and the comparison from original to equivalent codes has the same stubs to ensure parity
+> in the assembly that is being generated.
+
+## List of Documented Inline ASM
+
+---
+
+<details>
+<summary>GeneralsMD/Code/Libraries/Include/Lib/BaseType.h</summary>
+
+This file includes the following assembly blocks:
+
+<details>
+<summary>fast_float2long_round</summary>
+
+```c++
+__forceinline long fast_float2long_round(float f)
+{
+	long i;
+
+	__asm {
+		fld [f]
+		fistp [i]
+	}
+
+	return i;
+}
+```
+
+This is using assembly to do a fast round and conversion from `float` to `long`.
+
+My goto equivalent is:
+
+```c++
+#include <math.h>
+
+__froceinline float fast_float_trunc(float f) {
+    return lround(f);
```

Comment:
I can confirm that this does not produce the correct result.

Do we have any reference change from another repository that already fixes all inline assembler functions 100% correctly?